File: app/src/main/java/com/example/schoolairdroprefactoredition/scene/ssb/fragment/SSBViewModel.java
package com.example.schoolairdroprefactoredition.scene.ssb.fragment;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.example.schoolairdroprefactoredition.domain.DomainPostInfo;
import com.example.schoolairdroprefactoredition.domain.DomainPurchasing;
import com.example.schoolairdroprefactoredition.presenter.callback.ISSBCallback;
import com.example.schoolairdroprefactoredition.presenter.impl.SSBImpl;
import com.example.schoolairdroprefactoredition.scene.main.base.BaseStateViewModel;

public class SSBViewModel extends BaseStateViewModel implements ISSBCallback {

    private int sellingPage = 0;
    private int postsPage = 0;

    private final SSBImpl ssbImpl;

    private final MutableLiveData<DomainPurchasing> mSelling = new MutableLiveData<>();
    private final MutableLiveData<Boolean> mUnListItemResult = new MutableLiveData<>();
    private final MutableLiveData<Boolean> mUpdateItemResult = new MutableLiveData<>();

    private final MutableLiveData<DomainPostInfo> mPosts = new MutableLiveData<>();
    private final MutableLiveData<Boolean> mDeletePostResult = new MutableLiveData<>();
    private final MutableLiveData<Boolean> mUpdatePostResult = new MutableLiveData<>();

    private OnSSBActionListener mOnSSBActionListener;

    public SSBViewModel() {
        ssbImpl = SSBImpl.getInstance();
        ssbImpl.registerCallback(this);
    }

    public LiveData<DomainPurchasing> getSelling(int userID) {
        ssbImpl.getSellingList(userID);
        return mSelling;
    }

    public LiveData<Boolean> unListItem(String token, String goodsID) {
        ssbImpl.unListItem(token, goodsID);
        return mUnListItemResult;
    }

    public LiveData<DomainPurchasing> getSellingByID(int userID) {
        ssbImpl.getSellingByUID(userID);
        return mSelling;
    }

    @Override
    public void onSellingLoaded(DomainPurchasing selling) {
        mSelling.postValue(selling);
    }

    @Override
    public void onUnListItemSuccess() {
        mUnListItemResult.postValue(true);
    }

    @Override
    public void onActionFailed() {
        if (mOnSSBActionListener != null)
            mOnSSBActionListener.onActionFailed();
    }

    @Override
    public void onPostLoaded(DomainPostInfo postInfo) {
        mPosts.postValue(postInfo);
    }

    @Override
    public void onDeletePostSuccess() {
        mDeletePostResult.postValue(true);
    }

    @Override
    public void onError() {
        if (mOnRequestListener != null)
            mOnRequestListener.onError();
    }

    public interface OnSSBActionListener {
        void onActionFailed();
    }

    public void setOnSSBActionListener(OnSSBActionListener listener) {
        mOnSSBActionListener = listener;
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        ssbImpl.unregisterCallback(this);
    }
}

File: app/src/main/java/com/example/schoolairdroprefactoredition/viewmodel/SSBViewModel.kt
package com.example.schoolairdroprefactoredition.viewmodel

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.example.schoolairdroprefactoredition.domain.DomainIWant
import com.example.schoolairdroprefactoredition.domain.DomainSelling
import com.example.schoolairdroprefactoredition.domain.DomainUserIWant
import com.example.schoolairdroprefactoredition.repository.SSBRepository

class SSBViewModel : ViewModel() {

    /**
     * 当前获取的在售是第几页
     */
    private var sellingPage = 0

    /**
     * 当前获取的求购是第几页
     */
    private var iwantPage = 0

    private var userID = -1

    private val ssbRepository by lazy {
        SSBRepository.getInstance()
    }

    /**
     * 下架物品
     */
    fun deleteGoods(token: String, goodsID: String): LiveData<Boolean> {
        val deleteGoodsResultLiveData = MutableLiveData<Boolean>()
        ssbRepository.deleteGoods(token, goodsID) {
            deleteGoodsResultLiveData.postValue(it)
        }
        return deleteGoodsResultLiveData
    }

    /**
     * 获取用户在售物品
     */
    fun getSelling(userID: Int, isLoadMore: Boolean): LiveData<DomainSelling?> {
        val sellingLiveData = MutableLiveData<DomainSelling?>()
        if (isLoadMore) {
            sellingPage = 0
        }

        ssbRepository.getSelling(userID, ++sellingPage) {
            sellingLiveData.postValue(it)
        }
        return sellingLiveData
    }

    /**
     * 删除求购
     */
    fun deleteIWant(token: String, iwantID: String): LiveData<Boolean> {
        val deleteIWantResultLiveData = MutableLiveData<Boolean>()
        ssbRepository.deleteIWant(token, iwantID) {
            deleteIWantResultLiveData.postValue(it)
        }
        return deleteIWantResultLiveData
    }

    /**
     * 获取用户求购
     */
    fun getIWant(userID: Int, isLoadMore: Boolean): LiveData<DomainUserIWant?> {
        val iwantLiveData = MutableLiveData<DomainUserIWant?>()
        if (isLoadMore) {
            iwantPage = 0
        }
        ssbRepository.getIWant(userID, ++iwantPage) {
            iwantLiveData.postValue(it)
        }
        return iwantLiveData
    }
}